packages/unified-lint-rule/lib/index.js

```diff
@@ -18,10 +20,12 @@ import {wrap} from 'trough'
 const primitives = new Set(['string', 'number', 'boolean'])
 
 /**
- * @param {string} id
+ * @param {string|RuleMeta} meta
  * @param {Rule} rule
  */
-export function lintRule(id, rule) {
+export function lintRule(meta, rule) {
+  const id = typeof meta === 'string' ? meta : meta.origin
+  const url = typeof meta === 'string' ? undefined : meta.url
```

Comment:
I wonder if it would make sense to use some of the same semantics `eslint` uses? :thinking: 
They leverage a `docs` field https://eslint.org/docs/developer-guide/working-with-rules#rule-basics

> `docs` (object) is required for core rules of ESLint:
>
> *   `description` (string) provides the short description of the rule in the rules index
> *   `category` (string) specifies the heading under which the rule is listed in the rules index
> *   `recommended` (boolean) is whether the "extends": "eslint:recommended" property in a configuration file enables the rule
> *   `url` (string) specifies the URL at which the full documentation can be accessed (enabling code editors to provide a helpful link on highlighted rule violations)
> *   `suggestion` (boolean) specifies whether rules can return suggestions (defaults to false if omitted)

In particular, `description`, `category`, and `url` seem relevant.
And having it scoped (to `docs` or something else) keeps `url` open to any other uses we may have in the future.

Reply:
I could nest the `url` property inside an object property `docs`. Although none of the other properties seem to map to any properties relevant to vfile messages. I also don’t really see another use case for the `url` property.

I think it’s fine as-is, but if you feel strong about this and it’s ok with @wooorm,, I’ll introduce the `docs` object with only `url` for now.

Reply:
I don't feel strongly about scoping.
I do find `url` floating in meta confusing though.
If I saw
```js
{
  origin: "something",
  url: "https://example.com"
}
```
I'd have very little understanding of the intent.

Something like:
```js
{
  lintRuleName: "something",
  documentationUrl: "https://example.com"
}
```
would be more verbose, but clearer.

Reply:
I agree if you put it that way. Although [`url`](https://github.com/vfile/vfile-message#url) and [`origin`](https://github.com/vfile/vfile-message#origin) match the names of the properties on vfile messages they represent.

Reply:
Hmmm, thoughts @wooorm? :thought_balloon: 

Another approach could be to update the type definition to carry some description through.
https://github.com/remarkjs/remark-lint/pull/273#discussion_r704622023
Which would benefit folks using an IDE with TS-language-server support, but would still leave ambiguity otherwise.

Reply:
- `category` — that’s a bit like `origin`
- `recommended` — more needed because ESLint does stuff by default, whereas remark-lint is opt-in only
- `suggestion` — some retext rules add an `actual`/`expected`, but lint rules don’t
- `description` — how much should this be? Markdown? Short message? Isn’t either the error message or the full readme enough? `vfile-reporter` does support a `note` I believe btw, which can have some more information

---

Re Christian’s point on:

```json
{
  origin: "something",
  url: "https://example.com"
}
```

looking nondescript, I’d say that perhaps it is that way because of how it’s abstracted. A more practical example:

```json
{
  origin: 'remark-lint:blockquote-indentation',
  url: 'https://github.com/remarkjs/remark-lint/tree/main/packages/remark-lint-blockquote-indentation#readme'
},
```

looks clear to me.

---

And for types as an alternative, the description is already defined in the types (`@fileoverview`) and extracted and put into the readme.
We could:
- Put that description in the JSDoc attached to the rule
- Put the description in a string (and the URL), and instead generate the readme based on actual code rather than comments

Reply:
> We could:
> 
> - Put that description in the JSDoc attached to the rule
> - Put the description in a string (and the URL), and instead generate the readme based on actual code rather than comments

Yes, we could. Although I doubt this has much added value. `unified-lint-rule` could support `note` anyway, but I think this is out of scope for this PR.